README.md
[![Gitter](https://badges.gitter.im/redis-service-consul/Lobby.svg)](https://gitter.im/redis-service-consul/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge) [![Go Report Card](https://goreportcard.com/badge/github.com/seatgeek/resec)](https://goreportcard.com/report/github.com/seatgeek/resec) [![Build Status](https://travis-ci.org/seatgeek/resec.svg?branch=master)](https://travis-ci.org/seatgeek/resec)

<p align="center">
  <img src="https://s.gravatar.com/avatar/96b073f48aae741171d137f21c849d84?s=160" alt="Resec - Consul based highly available Redis replication agent" />
</p>

# Resec - Consul based highly available Redis replication agent

## Description

Resec is a successor to [Redis Sentinel](https://redis.io/topics/sentinel) and [redishappy](https://github.com/mdevilliers/redishappy) for handling high availability failover for Redis.

It avoids Redis Sentinel problems of remembering all the sentinels and all the redis servers that ever appeared in the replication cluster.

Resec master election is based on [Consul Locks](https://www.consul.io/docs/commands/lock.html) to provide single redis master instance.

Resec continuously monitors the status of redis instance and if it's alive, It starts 2 following processes:
* Monitor service of *master* for changes
    * if lock is not acquired, on every change of master it runs *SLAVE OF `Master.Address`*
* Trying to acquire lock to became master itself
    * once lock acquired it stops watching for master service changes
    * promotes redis to be *SLAVE OF NO ONE*

### Services and health checks
Resec registers service with [TTL](https://www.consul.io/docs/agent/checkshtml#TTL) health check with TTL twice as big as `HEALTHCHECK_INTERVAL` and updates consul every `HEALTHCHECK_INTERVAL` to maintain service in passing state

There are 2 options to work with services:
* Use `CONSUL_SERVICE_NAME` for tag based master/slave discovery
  * `MASTER_TAGS` must be provided for ability to watch master instance for changes.
* Use `CONSUL_SERVICE_PREFIX` for service name only based discovery
  * services in consul will look like `CONSUL_SERVICE_PREFIX`-`Replication.Role`

* If `ANNOUNCE_ADDR` is set it will be used for registration in consul, if it's not provided `REDIS_ADDR` will be used for registration in consul.
  * If `REDIS_ADDR` is localhost, only port will be announced to the consul.

### Redis Health
* If redis becomes unhealthy resec will stop the leader election. As soon as redis will become healthy again, resec will start the operation from the beginning.

## Usage

### Environment variables

Environment Variables |  Default       | Description
----------------------| ---------------| -------------------------------------------------
ANNOUNCE_ADDR         |                | IP:Port of Redis to be announced, by default service will be registered wi
CONSUL_SERVICE_NAME   |                | Consul service name for tag based service discovery
CONSUL_SERVICE_PREFIX | redis          | Name Prefix, will be followed by "-(master/slave)", ignored if CONSUL_SERVICE_NAME is used
CONSUL_LOCK_KEY       | resec/.lock    | KV lock location, should be overriden if multiple instances running in the same consul DC
CONSUL_LOCK_SESSION_NAME       | resec | Lock session Name to distinguish multiple resec masters on one host
CONSUL_LOCK_MONITOR_RETRIES | 3        | Number of retries of lock receives 500 Error from Consul
CONSUL_LOCK_MONITOR_RETRY_INTERVAL | 1s | Retry interval if lock receives 500 Error from Consul
CONSUL_DEREGISTER_SERVICE_AFTER | 72h  |
CONSUL_LOCK_TTL       | 15s            |
MASTER_TAGS           |                | Comma separated list of tags to be added to master instance. The first tag (index 0) is used to configure the role of the Redis/resec task, and *must* be different from index 0 in `SLAVE_TAGS`.
SLAVE_TAGS            |                | Comma separated list of tags to be added to slave instance. The first tag (index 0) is used to configure the role of the Redis/resec task, and *must* be different from index 0 in `MASTER_TAGS`.
HEALTHCHECK_INTERVAL  | 5s             |
HEALTHCHECK_TIMEOUT   | 2s             |
REDIS_ADDR            | 127.0.0.1:6379 |
REDIS_PASSWORD        |                |
LOG_LEVEL             | INFO           | Options are "DEBUG", "INFO", "WARN", "ERROR"

##### Environment variables to configure communication with consul are similar to [Consul CLI](https://www.consul.io/docs/commands/index.html#environment-variables)

### Permissions

Resec requires permissions for Consul in order to function correctly.
The Consul ACL token is passed as the environment variable `CONSUL_HTTP_TOKEN` .

#### Consul ACL Token Permissions

If the Consul cluster being used is running ACLs; the following ACL policy will allow Replicator the required access to perform all functions based on its default configuration:

```hcl
key "resec/" {
  policy = "write"
}
session "" {
  policy = "write"
}
service "" {
  policy = "write"
}
```

### Run the application

* with nomad:
```hcl
job "resec" {
  datacenters = ["dc1"]
  type        = "service"

  update {
    max_parallel = 1
    stagger      = "10s"
  }

  group "cache" {
    count = 3

    task "redis" {
      driver = "docker"
      config {
        image = "redis:alpine"
        command = "redis-server"
        args = [
          "/local/redis.conf"
        ]
        port_map {
          db = 6379
        }
      }
      // Let Redis know how much memory he can use not to be killed by OOM
      template {
        data = <<EORC
maxmemory {{ env "NOMAD_MEMORY_LIMIT" | parseInt | subtract 16 }}mb
EORC
        destination   = "local/redis.conf"
      }

      resources {
        cpu    = 500
        memory = 256
        network {
          mbits = 10
          port "db" {}
        }
      }
    }

    task "resec" {
      driver = "docker"
      config {
        image = "seatgeek/resec"
      }

      env {
        CONSUL_HTTP_ADDR = "http://${attr.unique.network.ip-address}:8500"
        REDIS_ADDR = "${NOMAD_ADDR_redis_db}"
      }

      resources {
        cpu    = 100
        memory = 64
        network {
          mbits = 10
        }
      }
    }
  }
}
```

* with docker-compose.yml:

```yaml
resec:
  image: seatgeek/resec
  environment:
    - CONSUL_HTTP_ADDR=1.2.3.4:8500
    - REDIS_ADDR=redis:6379
  container_name: resec
```

* with SystemD:
```
[Unit]
Description=resec - redis ha replication daemon
Requires=network-online.target
After=network-online.target

[Service]
EnvironmentFile=-/etc/default/resec
ExecStart=/usr/local/bin/resec
KillSignal=SIGQUIT
Restart=on-failure
RestartSec=5s

[Install]
WantedBy=multi-user.target

```

## Debugging

### Dump state via signal

If you send `USR1` signal to Resec, it will dump the state of the reconciler to `stdout` no matter the log level you started Resec with.

For exampel running `killall -USR1 resec` will yield logging output shown below

```text
(state.Consul) {
 Ready: (bool) true,
 Healthy: (bool) true,
 Master: (bool) true,
 MasterAddr: (string) "",
 MasterPort: (int) 0,
 Stopped: (bool) false
}
(state.Redis) {
 Healthy: (bool) false,
 Ready: (bool) false,
 Info: (state.RedisStatus) {
  Role: (string) "",
  Loading: (bool) false,
  MasterLinkUp: (bool) false,
  MasterLinkDownSince: (time.Duration) 0s,
  MasterSyncInProgress: (bool) false,
  MasterHost: (string) "",
  MasterPort: (int) 0
 },
 InfoString: (string) "",
 Stopped: (bool) false
}
```

if you send `USR2` signal to Resec, it will dump the full reconciler state + all related structs managed by the reconciler.

For example running `killall -USR2 resec` will yield logging output shown below

```text
(*reconciler.Reconciler)(0xc0000fc000)({
 consulCommandCh: (chan<- consul.Command) (cap=10) 0xc00001e660,
 consulState: (state.Consul) {
  Ready: (bool) true,
  Healthy: (bool) true,
  Master: (bool) true,
  MasterAddr: (string) "",
  MasterPort: (int) 0,
  Stopped: (bool) false
 },
 consulStateCh: (<-chan state.Consul) (cap=10) 0xc00001e720,
 forceReconcileInterval: (time.Duration) 2s,
 logger: (*logrus.Entry)(0xc00008c9b0)({
  Logger: (*logrus.Logger)(0xc00008c190)({
   Out: (*os.File)(0xc00000c020)({
    file: (*os.file)(0xc00001e180)({
     pfd: (poll.FD) {
      fdmu: (poll.fdMutex) {
       state: (uint64) 0,
       rsema: (uint32) 0,
       wsema: (uint32) 0
      },
      Sysfd: (int) 2,
      pd: (poll.pollDesc) {
       runtimeCtx: (uintptr) <nil>
      },
      iovecs: (*[]syscall.Iovec)(<nil>),
      csema: (uint32) 0,
      isBlocking: (uint32) 1,
      IsStream: (bool) true,
      ZeroReadIsEOF: (bool) true,
      isFile: (bool) true
     },
     name: (string) (len=11) "/dev/stderr",
     dirinfo: (*os.dirInfo)(<nil>),
     nonblock: (bool) false,
     stdoutOrErr: (bool) true
    })
   }),
   Hooks: (logrus.LevelHooks) {
   },
   Formatter: (*logrus.TextFormatter)(0xc0000710b0)({
    ForceColors: (bool) false,
    DisableColors: (bool) false,
    DisableTimestamp: (bool) false,
    FullTimestamp: (bool) true,
    TimestampFormat: (string) "",
    DisableSorting: (bool) false,
    QuoteEmptyFields: (bool) false,
    isTerminal: (bool) true,
    Once: (sync.Once) {
     m: (sync.Mutex) {
      state: (int32) 0,
      sema: (uint32) 0
     },
     done: (uint32) 1
    }
   }),
   Level: (logrus.Level) info,
   mu: (logrus.MutexWrap) {
    lock: (sync.Mutex) {
     state: (int32) 0,
     sema: (uint32) 0
    },
    disabled: (bool) false
   },
   entryPool: (sync.Pool) {
    noCopy: (sync.noCopy) {
    },
    local: (unsafe.Pointer) 0xc0000e0600,
    localSize: (uintptr) 0xc,
    New: (func() interface {}) <nil>
   }
  }),
  Data: (logrus.Fields) (len=1) {
   (string) (len=6) "system": (string) (len=10) "reconciler"
  },
  Time: (time.Time) 0001-01-01 00:00:00 +0000 UTC,
  Level: (logrus.Level) panic,
  Message: (string) "",
  Buffer: (*bytes.Buffer)(<nil>)
 }),
 reconcile: (bool) false,
 reconcileInterval: (time.Duration) 100ms,
 redisCommandCh: (chan<- redis.Command) (cap=10) 0xc00001e4e0,
 redisState: (state.Redis) {
  Healthy: (bool) false,
  Ready: (bool) false,
  Info: (state.RedisStatus) {
   Role: (string) "",
   Loading: (bool) false,
   MasterLinkUp: (bool) false,
   MasterLinkDownSince: (time.Duration) 0s,
   MasterSyncInProgress: (bool) false,
   MasterHost: (string) "",
   MasterPort: (int) 0
  },
  InfoString: (string) "",
  Stopped: (bool) false
 },
 redisStateCh: (<-chan state.Redis) (cap=10) 0xc00001e480,
 signalCh: (chan os.Signal) (cap=1) 0xc00001e840,
 debugSignalCh: (chan os.Signal) (cap=1) 0xc00001e900,
 stopCh: (chan interface {}) (cap=1) 0xc00001e8a0,
 Mutex: (sync.Mutex) {
  state: (int32) 0,
  sema: (uint32) 0
 }
})
```

### Tracking state changing via logging

If you run `resec` with `LOG_LEVEL=debug`, there will be a log line for each Consul/Redis state change the reconsider sees.

Example can be seen below

```text
DEBU[2019-01-01T16:18:01+01:00] New Consul state                              system=reconciler
DEBU[2019-01-01T16:18:01+01:00] modified: .Master = false                     system=reconciler
DEBU[2019-01-01T16:17:56+01:00] New Redis state                               system=reconciler
DEBU[2019-01-01T16:17:56+01:00] modified: .Healthy = true                     system=reconciler
DEBU[2019-01-01T16:17:56+01:00] New Redis state                               system=reconciler
DEBU[2019-01-01T16:17:56+01:00] modified: .Ready = true                       system=reconciler
DEBU[2019-01-01T16:17:56+01:00] modified: .Info.Role = "master"               system=reconciler
```

## Copyright and license

Code released under the [MIT license](https://github.com/seatgeek/ReSeC/blob/master/LICENSE).


CHANGELOG.md
# Changelog

## 0.6.0 (???)

- Upgrade Go to 1.13
- Use Go `mod` instead of `dep`
- Add `SIGUSR1` and `SIGUSR2` signal that will dump internal state for debugging purposes
- Add `LOG_LEVEL=debug` state diff output
- Fixed a gazillion typos of `reconciler`
- Split up the `eval` and `apply` in the reconciler for easier management
- Improve Redis and Consul manager logic and state management
- Make Redis connection handler do backoff like Consul
- Various code formatting improvements
- Build macOS and Linux binary on release

## 0.5.1 (January 16, 2019)

FIXES:

- Docker build

## 0.5.0 (September 16, 2018)

FEATURES:

- Configure via CLI arguments
- Add json/gelf log support
- Tests

FIXES:

- Split-brain when redis master restarted without restart of its resec

## 0.4.3 (July 15, 2018)

FIXES:

- Allow same Master and Slave tags when using CONSUL_SERVICE_PREFIX

## 0.4.2 (July 15, 2018)

FEATURES:

- Verbose Health Check names giving the option to distinguish checks of multiple Resec instances running on the same node

## 0.4.1 (May 16, 2018)

FEATURES:

- Prevent master and slave tag having index 0 identical when using CONSUL_SERVICE_NAME
- Test!

## 0.4.0 (April 26, 2018)

FEATURES:

- Service check now registered as initially passing to avoid service flapping

FIXES:

- Handling slave session invalidation
- Prevent from shutting down if local consul agent is not healthy

## 0.3.0 (April 23, 2018)

FEATURES:

- Consul lock session name can be adjusted
- Consul lock retry option added
- Dependencies updated (Consul to 1.0.7, go-redis to 6.10.2)

## 0.2.0 (March 13, 2018)

FEATURES:

- Consul service name for tag based service discovery
- Option to ADD (MASTER|SLAVE)_TAGS to be populated to consul
- Managing dependencies using [dep](https://github.com/golang/dep)
- Go version bumped to 0.10.x

## 0.1.0 (January 22, 2018)

- Initial release

FEATURES:

- Redis Health monitoring
- Consul Agent outage handling
- Looks working =)


init.go
package main

import (
	"fmt"
	"log"
	"net/http"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"syscall"
	"time"

	"github.com/go-redis/redis"
	consulapi "github.com/hashicorp/consul/api"
	"github.com/hashicorp/logutils"
)

// Environment variables keys
const (
	AnnounceAddr                   = "ANNOUNCE_ADDR"
	ConsulDeregisterServiceAfter   = "CONSUL_DEREGISTER_SERVICE_AFTER"
	ConsulLockKey                  = "CONSUL_LOCK_KEY"
	ConsulLockSessionName          = "CONSUL_LOCK_SESSION_NAME"
	ConsulLockTTL                  = "CONSUL_LOCK_TTL"
	ConsulLockMonitorRetries       = "CONSUL_LOCK_MONITOR_RETRIES"
	ConsulLockMonitorRetryInterval = "CONSUL_LOCK_MONITOR_RETRY_INTERVAL"
	ConsulServiceName              = "CONSUL_SERVICE_NAME"
	ConsulServicePrefix            = "CONSUL_SERVICE_PREFIX"
	HealthCheckInterval            = "HEALTHCHECK_INTERVAL"
	HealthCheckTimeout             = "HEALTHCHECK_TIMEOUT"
	LogLevel                       = "LOG_LEVEL"
	MasterTags                     = "MASTER_TAGS"
	RedisAddr                      = "REDIS_ADDR"
	RedisPassword                  = "REDIS_PASSWORD"
	SlaveTags                      = "SLAVE_TAGS"
)

// consul struct description
type consul struct {
	checkID                  string                 // consul check ID to register our service with
	client                   *consulapi.Client      // consul client used to interact with Consul
	clientConfig             *consulapi.Config      // consul client config
	deregisterServiceAfter   time.Duration          // time to have the consul service in "failed" mode before consul will reap it
	healthy                  bool                   // track if the consul backend is healthy or not
	lock                     *consulapi.Lock        // consul log struct
	lockAbortCh              chan struct{}          // channel will be closed when we want to abort waiting for the consul lock
	lockErrorCh              <-chan struct{}        // channel will be closed by Consul SDK when the held lock fails
	lockIsHeld               bool                   // track if we currently hold the lock (aka redis master)
	lockIsWaiting            bool                   // track if we are waiting for the lock to be released (trying to become master)
	lockKey                  string                 // consul kv path to the lock
	lockMonitorRetries       int                    // consul lock retries if 500 is received
	lockMonitorRetryInterval time.Duration          // consul lock retry time
	lockSessionName          string                 // consul lock session name
	lockStatusCh             chan *consulLockStatus // used to publish consul lock changes internally
	lockStopWaiterHandlerCh  chan bool              // used to gracefully release the held lock
	lockTTL                  time.Duration          // time-to-live for the lock (how frequently we need to update it to keep the ownership)
	lockWaitHandlerRunning   bool                   // track if we are currently waiting to acquire the lock
	serviceID                string                 // consul service ID (consul internal id) to use when registering the service
	serviceName              string                 // the (human) service name to register our service as
	serviceNamePrefix        string                 // an optional prefix for the consul service name
	tags                     map[string][]string    // list of arbitrary consul tags to publish with the service
	ttl                      string
}

// consulLockStatus is our internal representation of the consul lock health / avaibility
type consulLockStatus struct {
	acquired bool  // have we acquired the consul lock or not
	err      error // any errors we might have gotten during the consul lock
}

// redisConnection is our internal representation of redis state
type redisConnection struct {
	address           string        // address to connect to
	password          string        // (optional) redis password
	client            *redis.Client // redis client
	healthy           bool          // wether we can talk to redis or not
	replicationStatus string        // the current replication status (master or slave)
}

// redisInfo struct description
type redisInfo struct {
	address string
	port    int
}

// redisHealth struct description
type redisHealth struct {
	output  string // raw output from redis status
	healthy bool   // wether we successfully could talk to redis or not
}

// resec core struct
// holds references to other structs with internal state
type resec struct {
	announceAddr          string                         // what IP we should publish to Consul (default "")
	announceHost          string                         // ??
	announcePort          int                            // what Port we should publish to Consul (default "")
	consul                *consul                        // internal Consul state
	healthCheckInterval   time.Duration                  // how frequently we should update our consul check
	healthCheckTimeout    time.Duration                  // how long time the health check is allowed to take
	lastKnownMaster       *consulapi.ServiceEntry        // struct with the info on the last known redis master (from consul catalog)
	lastKnownMasterInfo   redisInfo                      // struct with the info on the last known redis master (internal representation)
	logLevel              string                         // which level we should do logging
	consulMasterServiceCh chan []*consulapi.ServiceEntry // channel where consul updates to the master service will be published
	redis                 *redisConnection               // internal Redis state
	redisReplicationCh    chan *redisHealth              // channel where redis updates will be published
	sigCh                 chan os.Signal                 // signal handler channel
	stopCh                chan interface{}               // stop channel for non-signal termination
}

// setup returns the default configuration for the ReSeC
func setup() (*resec, error) {
	config := &resec{
		consul: &consul{
			clientConfig: &consulapi.Config{
				HttpClient: &http.Client{
					Timeout: time.Second * 1,
				},
			},
			deregisterServiceAfter:   time.Hour * 72,
			lockAbortCh:              make(chan struct{}, 1),
			lockKey:                  "resec/.lock",
			lockMonitorRetries:       3,
			lockMonitorRetryInterval: time.Second * 1,
			lockSessionName:          "resec",
			lockStatusCh:             make(chan *consulLockStatus, 1),
			lockTTL:                  time.Second * 15,
			tags:                     make(map[string][]string),
			serviceNamePrefix:        "redis",
		},
		redis: &redisConnection{
			address: "127.0.0.1:6379",
		},
		logLevel:              "INFO",
		healthCheckInterval:   time.Second * 5,
		healthCheckTimeout:    2 * time.Minute,
		consulMasterServiceCh: make(chan []*consulapi.ServiceEntry, 1),
		redisReplicationCh:    make(chan *redisHealth, 1),
		sigCh:                 make(chan os.Signal, 1),
		stopCh:                make(chan interface{}, 1),
	}

	signal.Notify(config.sigCh, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGINT)

	if logLevel := os.Getenv(LogLevel); logLevel != "" {
		config.logLevel = logLevel
	}

	filter := &logutils.LevelFilter{
		Levels:   []logutils.LogLevel{"DEBUG", "INFO", "WARN", "ERROR", "CRITICAL"},
		MinLevel: logutils.LogLevel(config.logLevel),
		Writer:   os.Stderr,
	}
	log.SetOutput(filter)

	// Prefer CONSUL_SERVICE_NAME over CONSUL_SERVICE_PREFIX
	if consulServiceName := os.Getenv(ConsulServiceName); consulServiceName != "" {
		config.consul.serviceName = consulServiceName
	} else if consulServicePrefix := os.Getenv(ConsulServicePrefix); consulServicePrefix != "" {
		config.consul.serviceNamePrefix = consulServicePrefix
	}

	// Fail if CONSUL_SERVICE_NAME is used and no MASTER_TAGS are provided
	if masterTags := os.Getenv(MasterTags); masterTags != "" {
		config.consul.tags["master"] = strings.Split(masterTags, ",")
	} else if config.consul.serviceName != "" {
		return nil, fmt.Errorf("[ERROR] MASTER_TAGS is required when CONSUL_SERVICE_NAME is used")
	}

	if slaveTags := os.Getenv(SlaveTags); slaveTags != "" {
		config.consul.tags["slave"] = strings.Split(slaveTags, ",")
	}

	if config.consul.serviceName != "" {
		if len(config.consul.tags["slave"]) >= 1 && len(config.consul.tags["master"]) >= 1 {
			if config.consul.tags["slave"][0] == config.consul.tags["master"][0] {
				return nil, fmt.Errorf("[PANIC] The first tag in %s and %s must be unique", MasterTags, SlaveTags)
			}
		}
	}

	if consulLockKey := os.Getenv(ConsulLockKey); consulLockKey != "" {
		config.consul.lockKey = consulLockKey
	}

	if consulLockSessionName := os.Getenv(ConsulLockSessionName); consulLockSessionName != "" {
		config.consul.lockSessionName = consulLockSessionName
	}

	if consulLockMonitorRetries := os.Getenv(ConsulLockMonitorRetries); consulLockMonitorRetries != "" {
		consulLockMonitorRetriesInt, err := strconv.Atoi(consulLockMonitorRetries)
		if err != nil {
			return nil, fmt.Errorf("[ERROR] Trouble parsing %s [%s] as int: %s", ConsulLockMonitorRetries, consulLockMonitorRetries, err)
		}

		config.consul.lockMonitorRetries = consulLockMonitorRetriesInt
	}

	if consulLockMonitorRetryInterval := os.Getenv(ConsulLockMonitorRetryInterval); consulLockMonitorRetryInterval != "" {
		consulLockMonitorRetryIntervalDuration, err := time.ParseDuration(consulLockMonitorRetryInterval)
		if err != nil {
			return nil, fmt.Errorf("[ERROR] Trouble parsing %s [%s] as time: %s", ConsulLockMonitorRetryInterval, consulLockMonitorRetryInterval, err)
		}

		config.consul.lockMonitorRetryInterval = consulLockMonitorRetryIntervalDuration
	}

	if healthCheckInterval := os.Getenv(HealthCheckInterval); healthCheckInterval != "" {
		healthCheckIntervalDuration, err := time.ParseDuration(healthCheckInterval)
		if err != nil {
			return nil, fmt.Errorf("[ERROR] Trouble parsing %s [%s] as time: %s", HealthCheckInterval, healthCheckInterval, err)
		}

		config.healthCheckInterval = healthCheckIntervalDuration
	}

	// setting Consul Check TTL to be 2 * Check Interval
	config.consul.ttl = time.Duration(config.healthCheckInterval * 2).String()

	if healthCheckTimeout := os.Getenv(HealthCheckTimeout); healthCheckTimeout != "" {
		healthCheckTimeOutDuration, err := time.ParseDuration(healthCheckTimeout)
		if err != nil {
			return nil, fmt.Errorf("[ERROR] Trouble parsing %s [%s] as time: %s", HealthCheckTimeout, healthCheckTimeout, err)
		}

		config.healthCheckTimeout = healthCheckTimeOutDuration
	}

	if consulDeregisterServiceAfter := os.Getenv(ConsulDeregisterServiceAfter); consulDeregisterServiceAfter != "" {
		consulDeregisterServiceAfterDuration, err := time.ParseDuration(consulDeregisterServiceAfter)
		if err != nil {
			return nil, fmt.Errorf("[ERROR] Trouble parsing %s [%s] as time: %s", ConsulDeregisterServiceAfter, consulDeregisterServiceAfter, err)
		}

		config.consul.deregisterServiceAfter = consulDeregisterServiceAfterDuration

	}

	if consuLockTTL := os.Getenv(ConsulLockTTL); consuLockTTL != "" {
		consuLockTTLDuration, err := time.ParseDuration(consuLockTTL)
		if err != nil {
			return nil, fmt.Errorf("[ERROR] Trouble parsing %s [%s] as time: %s", ConsulLockTTL, consuLockTTL, err)
		}

		if consuLockTTLDuration < time.Second*15 {
			return nil, fmt.Errorf("[CRITICAL] Minimum Consul lock session TTL is 15s")
		}

		config.consul.lockTTL = consuLockTTLDuration
	}

	if redisAddr := os.Getenv(RedisAddr); redisAddr != "" {
		config.redis.address = redisAddr
	}

	if redisPassword := os.Getenv(RedisPassword); redisPassword != "" {
		config.redis.password = redisPassword
	}

	// If ANNOUNCE_ADDR is set it will be used for registration in consul
	// otherwise if redis address is provided - it will be used for registration in consul
	// if redis address is localhost only port will be announced to the consul

	if announceAddr := os.Getenv(AnnounceAddr); announceAddr != "" {
		config.announceAddr = announceAddr
	} else {
		redisHost := strings.Split(config.redis.address, ":")[0]
		redisPort := strings.Split(config.redis.address, ":")[1]
		if redisHost == "127.0.0.1" || redisHost == "localhost" || redisHost == "::1" {
			config.announceAddr = ":" + redisPort
		} else {
			config.announceAddr = config.redis.address
		}
	}

	var err error
	config.announceHost = strings.Split(config.announceAddr, ":")[0]
	config.announcePort, err = strconv.Atoi(strings.Split(config.announceAddr, ":")[1])
	if err != nil {
		return nil, fmt.Errorf("[ERROR] Trouble extracting port number from [%s]", config.redis.address)
	}

	// initialise redis as unhealthy
	config.redis.healthy = false

	redisOptions := &redis.Options{
		Addr:        config.redis.address,
		DialTimeout: config.healthCheckTimeout,
		ReadTimeout: config.healthCheckTimeout,
	}

	if config.redis.password != "" {
		redisOptions.Password = config.redis.password
	}

	config.redis.client = redis.NewClient(redisOptions)

	config.consul.client, err = consulapi.NewClient(config.consul.clientConfig)

	if err != nil {
		return nil, fmt.Errorf("[CRITICAL] Can't initialize consul client %s", err)
	}

	return config, nil
}
